aiida_vasp/workflows/relax.py

```diff
@@ -0,0 +1,298 @@
+"""Structure relaxation workchain for VASP."""
+import numpy
+from aiida.common.extendeddicts import AttributeDict
+from aiida.work import WorkChain
+from aiida.work.workchain import append_, while_, ToContext
+
+from aiida_vasp.utils.aiida_utils import get_data_class, get_data_node
+from aiida_vasp.utils.vasp.isif import IsifStressFlags, Isif
+from aiida_vasp.utils.vasp.ibrion import IbrionFlags, Ibrion
+from aiida_vasp.utils.vasp.encut import EncutFlags, Encut
+from aiida_vasp.utils.vasp.nsw import Nsw
+from aiida_vasp.workflows.base import VaspBaseWf
+from aiida_vasp.workflows.restart import prepare_process_inputs, UnexpectedCalculationFailure
+
+RELAXATION_INCAR_TEMPLATE = AttributeDict({
+    "ismear": -1,
+    "sigma": 0.2,
+    "algo": 'NORMAL',
+    "ediff": 1E-6,
+    "prec": 'Normal',
+    "ibrion": 2,
+    "nsw": 200,
+    "nelmin": 4,
+    "ediffg": -0.01,
+    "isif": 4
+})
+
+
+def relax_parameters(pos, shape, volume):
+    """Set IBRION and ISIF according to the choice of what should be relaxed."""
+    if not (pos or volume or shape):
+        raise ValueError('relaxation workflow requires at least one of "positions", "cell shape", "cell volume" to be alowed to change.')
+    if not pos:
+        ibrion = Ibrion(ion_updates=IbrionFlags.NO_UPDATE)
+    else:
+        ibrion = Ibrion(ion_updates=IbrionFlags.IONIC_RELAXATION_CG)
+    isif = Isif(calculate_stress=IsifStressFlags.FULL, vary_positions=pos, vary_cell_shape=shape, vary_cell_volume=volume)
+    nsw = Nsw(value=200, ibrion=ibrion)
+    return ibrion, isif, nsw
+
+
+def encut_factor(volume):
+    if volume:
+        return 1.3
+    return 1
+
+
+def clean_incar_overrides(inputs):
+    """Make sure the incar overrides do not clash with parameters necessary for relaxation."""
+    overrides = AttributeDict(inputs.incar_add.get_dict())
+    if 'ibrion' in overrides:
+        raise ValueError('overriding IBRION not allowed, use relax.xxx inputs to control')
+    if 'isif' in overrides:
+        raise ValueError('overriding ISIF not allowed, use relax.xxx inputs to control')
+    if 'nsw' in overrides:
+        if inputs.relax.positions.value and overrides.nsw < 1:
+            raise ValueError('NSW (num ionic steps) was set to 0 but relaxing positions was requested')
+
+    return overrides
+
+
+def l2_norm(vector_a):
```

Comment:
Why not `np.linalg.norm`?

Reply:
Ah, that's the one I was looking for. Didn't have access to the docs when I wrote that...